### PR TITLE
[OpenAPI3 Emitter] Add option to map `safeint` to `double-int`

### LINKED ISSUE #2367: [OpenAPI] Safeint emits int64 format
While there may not be an accurate `format` specifier for safeint, I believe the current emit of `format: int64` feels incorrect. The entire point of safeint is an int that fits safely into a double, so the ideal JS emit for this would indeed be a double. But the emitted OpenAPI doesn't indicate this, so a conservative emitter would pick BigInteger (at severe ergonomic cost).

I believe we have a few options, any of which would be better IMO:

1. argue for adding `int53` to the format registry and use that if accepted
2. just use `int53`
3. use `double`
4. omit format entirely

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix  #2367 
Format was added to the schema registry so we are safe to use that now https://spec.openapis.org/registry/format/double-int.html

### CODE REVIEW

File: .chronus/changes/safe-int-openapi3-2024-1-17-0-53-2.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+---
+# Change versionKind to one of: internal, fix, dependencies, feature, deprecation, breaking
+changeKind: breaking
+packages:
+  - "@typespec/openapi3"
+---
+
+`safeint` scalar now map to `type: integer, format: double-int`. To revert to the previous output of `type: integer, format: int64`. You can either:
```
Comment (bterlson) [inline .chronus/changes/safe-int-openapi3-2024-1-17-0-53-2.md:8]:
```suggestion
The `safeint` scalar now emits to `type: integer, format: double-int`. The `double-int` format was added to the [OpenAPI Formats Registry](https://spec.openapis.org/registry/format/double-int) in November 2023 to represent integers that can be stored in a double without precision loss.

There are two options to preserve the previous emit of `type: integer, format: int64`:
```

File: packages/openapi3/README.md [hunk 2]
```diff
@@ -86,6 +86,15 @@ By default all types declared under the service namespace will be included. With
...
 This extension is meant for debugging and should not be depended on.
 
+#### `safeint-strategy`
+
+**Type:** `"double-int" | "int64"`
+
+How to handle safeint type. Options are:
+
+- `double-int`: Will produce `type: integer, format: double-int`
+- `int64`: Will produce `type: integer, format: int64`
```
Comment (markcowl) [inline packages/openapi3/README.md:96]: It's a shame the default setting isn't generated here

File: packages/openapi3/src/lib.ts [hunk 3]
```diff
@@ -54,6 +54,14 @@ export interface OpenAPI3EmitterOptions {
    * @default "never"
    */
   "include-x-typespec-name"?: "inline-only" | "never";
+
+  /**
+   * How to handle safeint type. Options are:
```
Comment (markcowl) [inline packages/openapi3/src/lib.ts:59]: should likely add a blurb about the default:  if not specified, 'int64' will be used.

File: packages/openapi3/src/lib.ts [hunk 4]
```diff
@@ -117,6 +125,17 @@ const EmitterOptionsSchema: JSONSchemaType<OpenAPI3EmitterOptions> = {
       description:
         "If the generated openapi types should have the `x-typespec-name` extension set with the name of the TypeSpec type that created it.\nThis extension is meant for debugging and should not be depended on.",
     },
+    "safeint-strategy": {
+      type: "string",
+      enum: ["double-int", "int64"],
+      nullable: true,
+      default: "int64",
+      description: [
+        "How to handle safeint type. Options are:",
```
Comment (markcowl) [inline packages/openapi3/src/lib.ts:134]: same.  Is this auto-generated?

File: packages/openapi3/src/schema-emitter.ts [hunk 5]
```diff
@@ -699,7 +699,13 @@ export class OpenAPI3SchemaEmitter extends TypeEmitter<
       case "int64":
         return { type: "integer", format: "int64" };
       case "safeint":
-        return { type: "integer", format: "int64" };
+        switch (this.#options.safeintStrategy) {
+          case "int64":
+            return { type: "integer", format: "int64" };
+          case "double-int":
+          default:
+            return { type: "integer", format: "double-int" };
```
Comment (markcowl) [inline packages/openapi3/src/schema-emitter.ts:707]: int64 should be the default

File: packages/openapi3/test/primitive-types.test.ts [hunk 6]
```diff
@@ -48,6 +48,19 @@ describe("openapi3: primitives", () => {
     }
   });
 
+  it("produce type: integer, format: double-int for safeint when safeint-strategy is double-int", async () => {
```
Comment (markcowl) [inline packages/openapi3/test/primitive-types.test.ts:51]: Need to add a test to ensure that the value is base64 when no value is specified (do we have this already)?  Also, should ensure we get base64 when this is officially set.

Comment (markcowl) [reply packages/openapi3/test/primitive-types.test.ts:51]: Also, we should have a test in which an `@encode` is added, which I think should win out over this setting

**Review by markcowl — CHANGES_REQUESTED** [review]

File: packages/openapi3/test/primitive-types.test.ts (hunk 6 above)
Comment (timotheeguerin) [reply packages/openapi3/test/primitive-types.test.ts:51]: you mean int64? we do have the test above that deal with the default

Comment (timotheeguerin) [reply packages/openapi3/test/primitive-types.test.ts:51]: I can add an explicit with the option as well

File: packages/openapi3/src/lib.ts (hunk 4 above)
Comment (timotheeguerin) [reply packages/openapi3/src/lib.ts:134]: This one has to be written manually until we have the config as tsp

File: packages/openapi3/README.md (hunk 2 above)
Comment (timotheeguerin) [reply packages/openapi3/README.md:96]: yeah but its something we could add to the ref doc emitter.

**Review by markcowl — APPROVED** [review]

**Pull request merged**